Edits for tone and style (#23)
uptane/deployment-considerations: customizations.md

## Delta update strategies

In order to save bandwidth costs, Uptane allows an OEM to deliver updates as

> delta images. A delta image update contains only the code and/or data that differs from the image installed by the ECU. In order to use delta images, the OEM SHOULD make the following changes.

Author: Thanks for picking up on the extra space on the "and/or"s. I've been trying to fix these as I see them. Guess I missed this one. The other suggested change is fine.

> delta images. A delta image update contains only the code and/or data that differs from the image installed by the ECU. In order to use delta images, the OEM SHOULD make the following changes.

Reviewer: Nah, you didn't miss this one, I just had additional suggestions in addition to that. :)

**Figure 2.** *Two options for fleet management with Uptane.*

Some parties, such as vehicle rental companies or the military, may wish to exercise control on how their own fleet of vehicles are updated. Uptane offers two options for serving these users, as illustrated in Figure 2. Choosing between them depends on whether the fleet manager wishes to have either complete control, or better compromise-resilience.

> In the first option, which we expect to be the common case, a fleet manager would configure the map file on ECUs such that Primaries and full verification Secondaries would trust an image if, and only if, it has been signed by both the OEM-managed Image repository, and the fleet-managed Director repository. Partial verification Secondaries would trust an image, if and only, if it has been signed by this Director repository. The upside of this option is that the fleet manager, instead of the OEM, has complete control over which updates are installed on its vehicles. The downside of this option is that should the fleet-managed Directory repository be compromised, attackers can execute mix-and-match attacks.

Author: I added those commas on the last go-ahead for emphasis. That is to add some weight to the decision. If the commas are distracting, maybe we should italicize only. I think we want that "and only if" phrase to stand out. See if below is better...

> In the first option, which we expect to be the common case, a fleet manager would configure the map file on ECUs such that Primaries and full verification Secondaries would trust an image if, and only if, it has been signed by both the OEM-managed Image repository, and the fleet-managed Director repository. Partial verification Secondaries would trust an image, if and only, if it has been signed by this Director repository. The upside of this option is that the fleet manager, instead of the OEM, has complete control over which updates are installed on its vehicles. The downside of this option is that should the fleet-managed Directory repository be compromised, attackers can execute mix-and-match attacks.

Reviewer: I like the rephrasing with the key word "only". But I wonder if it would be better to emphasize the word "both" instead; I think that's now the critical part.

How did you feel about my change to the last sentence? I found the "should ... be" construction just a bit harder to understand than "if ... is". So if I combine all of that, I'd end up with:
```suggestion
In the first option, which we expect to be the common case, a fleet manager would configure the map file on ECUs such that Primaries and full verification Secondaries would only trust an image that has been signed by *both* the OEM-managed Image repository and the fleet-managed Director repository. Partial verification Secondaries would only trust an image if it has been signed by the fleet-managed Director repository. The upside of this option is that the fleet manager, instead of the OEM, has complete control over which updates are installed on its vehicles. The downside of this option is that if the fleet-managed Directory repository is compromised, attackers can execute mix-and-match attacks.
```

> In the second option, a fleet manager would configure the map file on ECUs such that Primaries and full verification Secondaries would trust an image if, and only if, it has been signed by three repositories: the OEM-managed Image repository, the OEM-managed Director repository, and the fleet-managed Director repository. The upside of this option is that attackers cannot execute mix-and-match attacks if they have compromised only one of the Director repositories. The downside of this option is that updates cannot be installed on vehicles unless both the OEM and fleet agree on which images should be installed. This agreement may require both Director repositories to communicate using an out-of-band channel. Using this option also means that partial verification Secondaries should be configured to trust the Director repository managed by either the OEM or the fleet, but not both, since these Secondaries can only check for one signature.

Author: I would use the same alternate suggested text above.
```suggestion
In the second option, a fleet manager would configure the map file on ECUs such that Primaries and full verification Secondaries would *only* trust an image that has been signed by three repositories: the OEM-managed Image repository, the OEM-managed Director repository, and the fleet-managed Director repository. The upside of this option is that attackers cannot execute mix-and-match attacks if they have compromised only one of the Director repositories. The downside of this option is that updates cannot be installed on vehicles unless both the OEM and fleet agree on which images should be installed. This agreement may require both Director repositories to communicate using an out-of-band channel. Using this option also means that partial verification Secondaries should be configured to either trust the Director repository managed by the OEM or by the fleet, but not both, since these Secondaries can only check for one signature.
```

> In the second option, a fleet manager would configure the map file on ECUs such that Primaries and full verification Secondaries would trust an image if, and only if, it has been signed by three repositories: the OEM-managed Image repository, the OEM-managed Director repository, and the fleet-managed Director repository. The upside of this option is that attackers cannot execute mix-and-match attacks if they have compromised only one of the Director repositories. The downside of this option is that updates cannot be installed on vehicles unless both the OEM and fleet agree on which images should be installed. This agreement may require both Director repositories to communicate using an out-of-band channel. Using this option also means that partial verification Secondaries should be configured to trust the Director repository managed by either the OEM or the fleet, but not both, since these Secondaries can only check for one signature.

Reviewer: This is fine, or as above, instead of emphasizing "only", emphasize "three".

> In the second option, a fleet manager would configure the map file on ECUs such that Primaries and full verification Secondaries would trust an image if, and only if, it has been signed by three repositories: the OEM-managed Image repository, the OEM-managed Director repository, and the fleet-managed Director repository. The upside of this option is that attackers cannot execute mix-and-match attacks if they have compromised only one of the Director repositories. The downside of this option is that updates cannot be installed on vehicles unless both the OEM and fleet agree on which images should be installed. This agreement may require both Director repositories to communicate using an out-of-band channel. Using this option also means that partial verification Secondaries should be configured to trust the Director repository managed by either the OEM or the fleet, but not both, since these Secondaries can only check for one signature.

Author: Ok. I'll be fixing that in a follow-up commit that fixes the wording of "map file"

## Custom installation instructions for ECUs

Most inputs to ECUs are delivered as signed Targets files, stored on the Image directory, and then sent to the ECU by the Director. However, there may be some cases where the inputs required for a particular customization cannot be configured to follow this standard signing process. Variations in input may be due to not knowing the input in advance, or a need to customize instructions for each vehicle. Examples of such inputs could be a command line option that turns on a feature in certain ECUs, a configuration sent by a Director repository to an ECU, or a Director doing a dynamic customization for an ECU. We can collectively call all these non-standard inputs "dynamic directions." Uptane allows ECUs to access dynamic directions in two different ways, each having particular advantages for different use cases.

### Accessing dynamic directions through signed images from the Director repository

> The first option for providing dynamic directions is to slightly modify the standard delivery procedure described above. An  image signed by the Director, Timestamp, and Metadata roles would still be sent to the ECU from the Director repository, but this file would not be stored on—-or validated by—-the Image repository. As the Image repository is controlled by offline keys, it can not validate a file created dynamically by the Director.

Author: Good question. Maybe we just pull out those references.
```suggestion
The first option for providing dynamic directions is to slightly modify the standard delivery procedure described above. The Director repository would still send a signed image to the ECU, but this file would not be stored on—-or validated by—-the Image repository. As the Image repository is controlled by offline keys, it can not validate a file created dynamically by the Director.
```

> The first option for providing dynamic directions is to slightly modify the standard delivery procedure described above. An  image signed by the Director, Timestamp, and Metadata roles would still be sent to the ECU from the Director repository, but this file would not be stored on—-or validated by—-the Image repository. As the Image repository is controlled by offline keys, it can not validate a file created dynamically by the Director.

Reviewer: Your suggestion looks good to me, thanks!

Even though the Image repository can not sign the file, this modification still provides some security protections. The ECU would continue to have rollback protection for a file sent this way, as a release counter will still be included in the metadata, and incremented for each new version. If additional validation is needed, the file could be put on multiple repositories created for this purpose. These repositories could behave similar to the Director repository, but would all have separate keys to allow for additional security. The Primary ECU will be aware of these extra repositories so it can check for consistency by downloading and comparing the image from all repositories.

### Adding dynamic directions to the custom field of Targets metadata

> Another way to provide dynamic directions is to use the custom field of the Targets metadata file. This field provides the option to include custom inputs to individual ECUs. Using the custom field is an especially good option for managing small variations in the existing image. For example, a compilation flag to enable a navigation feature might be set on some ECUs, but not on others. The custom field could contain dynamic directions, and additional subfields would help determine for which ECUs the direction is intended. In the flag example above, the Director can put the ECU id and the flag into the custom field so the flag will be used during the installation process only on that particular ECU. This custom field can then be included in the Targets metadata received by all ECUs. The intended ECU would be able to check for this flag and use it during an installation or update to enable the navigation system.

Author: Yes. Good catch.

> Another way to provide dynamic directions is to use the custom field of the Targets metadata file. This field provides the option to include custom inputs to individual ECUs. Using the custom field is an especially good option for managing small variations in the existing image. For example, a compilation flag to enable a navigation feature might be set on some ECUs, but not on others. The custom field could contain dynamic directions, and additional subfields would help determine for which ECUs the direction is intended. In the flag example above, the Director can put the ECU id and the flag into the custom field so the flag will be used during the installation process only on that particular ECU. This custom field can then be included in the Targets metadata received by all ECUs. The intended ECU would be able to check for this flag and use it during an installation or update to enable the navigation system.

Reviewer: Yep, looks good.

> Using this method of providing dynamic directions offers several ways to secure the system. The Targets metadata is created by the Director, validated using the Timestamp, and stored in the Image repository. As an added protection against a compromise of the Director role, the custom field could also be included in the Targets metadata file on the Image repository. However, if there are multiple customization possibilities, the better option would be to store the Targets metadata without the custom field in the Image repository, and keep the custom field configured separately and signed by the Director.

Author: Ok. This is one I can't answer as this text came right from the original document. @trishankkarthik @JustinCappos  can one of of address this question?

> Using this method of providing dynamic directions offers several ways to secure the system. The Targets metadata is created by the Director, validated using the Timestamp, and stored in the Image repository. As an added protection against a compromise of the Director role, the custom field could also be included in the Targets metadata file on the Image repository. However, if there are multiple customization possibilities, the better option would be to store the Targets metadata without the custom field in the Image repository, and keep the custom field configured separately and signed by the Director.

Reviewer 2: I don't remember writing this paragraph, and I think it has mistakes. Let me edit it to say what I think was meant.

> Using this method of providing dynamic directions offers several ways to secure the system. The Targets metadata is created by the Director, validated using the Timestamp, and stored in the Image repository. As an added protection against a compromise of the Director role, the custom field could also be included in the Targets metadata file on the Image repository. However, if there are multiple customization possibilities, the better option would be to store the Targets metadata without the custom field in the Image repository, and keep the custom field configured separately and signed by the Director.

Author: Trishank, I think this was added to Deployment. It was not part of your original text. If you can clarify better, please do so. And, thanks.
Lois

> However, note that using this method of providing dynamic directions means that a compromise of the Director repository might be able to cause ECUs to misconfigure their images. One way to mitigate this risk would be to require the Image repository to sign off on exactly the same directions using its own custom Targets metadata. but this is difficult to achieve considering that such directions are supposed to be dynamic in the first place. Therefore, proceed, if necessary, with caution.

Reviewer 2: ```suggestion
However, note that using this method of providing dynamic directions means that a compromise of the Director repository might be able to cause ECUs to misconfigure their images. One way to mitigate this risk would be to require the Image repository to sign off on exactly the same directions using its own custom Targets metadata. However, this is difficult to achieve considering that such directions are supposed to be dynamic in the first place. Therefore, proceed, if necessary, with caution.
```

Author: Looks good to me. I'm going to commit and merge.

> Using this method of providing dynamic directions offers several ways to secure the system. The Targets metadata is created by the Director, validated using the Timestamp, and stored in the Image repository. As an added protection against a compromise of the Director role, the custom field could also be included in the Targets metadata file on the Image repository. However, if there are multiple customization possibilities, the better option would be to store the Targets metadata without the custom field in the Image repository, and keep the custom field configured separately and signed by the Director.

Reviewer 2: ```suggestion
However, note that using this method of providing dynamic directions means that a compromise of the Director repository might be able to cause ECUs to misconfigure their images. One way to mitigate this risk would be to require the Image repository to sign off on exactly the same directions using its own custom targets metadata, but this is difficult to achieve considering that such directions are supposed to be dynamic in the first place. Therefore, proceed, if necessary, with caution.
```

Reviewer 2: @patrickvacek @jhdalek55 I added my suggested text. Thoughts?

Author: I accepted the suggestion with a small change. @patrickvacek does this edit work for you?
